docs/custom-plugins/custom-decorators.md

```diff
@@ -0,0 +1,158 @@
+
+To use this decorator, add it to a plugin. In this example the main decorator file is `plugins/sparkle.js`:
+
+```js
+const OperationSparkle = require('./decorators/operation-sparkle.js');
+
+module.exports = {
+  id: 'sparkle',
+  decorators: {
+    oas3: {
+      'operation-sparkle': OperationSparkle,
+    }
+  }
+};
+```
+
+The plugin is good to go. For a user to include it in their Redocly configuration, edit the configuration file to look something like this:
+
+```yaml
+plugins:
+  - plugins/sparkle.js
+
+decorators:
+  sparkle/operation-sparkle: on
+```
+
+## Decorator example with parameters
+
+A common use case is a decorator that can accept input values to be used during processing. This example decorator adds a suffix to all OperationIds in the document. Since every use case is different, the user can configure what should be used for the `suffix` value.
+
+Here's the decorator code, in a file named `plugins/decorations/add-suffix.js` and it expects a configuration option named `suffix`:
+
+```js
+module.exports = OpIdSuffix;
+
+function OpIdSuffix({suffix}) {
+  console.log("updating OperationIds ... ");
+  return {
+    Operation: {
+      leave(target) {
+        if(target.operationId) {
+          target.operationId = target.operationId + suffix;
+        }
+      }
+    },
+  }
+};
+```
+
+The `suffix` configuration option is automatically passed in, and it can be used in the function.
+
+Now extend the decorator from the previous example to add this to the existing plugin in `plugins/sparkle.js`:
+
+```js
+const OperationSparkle = require('./decorators/operation-sparkle.js');
+const OpIdSuffix = require('./decorators/add-suffix.js');
+
+module.exports = {
+  id: 'sparkle',
+  decorators: {
+    oas3: {
+      'operation-sparkle': OperationSparkle,
+      'add-opid-suffix': OpIdSuffix,
+    }
+  }
+};
+```
+
+All that remains is for a user to configure this decorator in their `redocly.yaml` configuration file to take advantage of the new decorator functionality. Here's an example of the configuration file:
+
+```yaml
+plugins:
+  - plugins/sparkle.js
+
+decorators:
+  sparkle/operation-sparkle: on
+  sparkle/add-opid-suffix:
+    suffix: ButShinier
+```
+
+With this configuration, an `operationId` called `GetAllItems` would be rewritten as `GetAllItemsButShinier`. You can choose a more sensible suffix for your use case as appropriate.
+
+## Further examples of custom decorators
+
+See some more examples of decorators:
+- Follow our [replace-servers-url tutorial](../guides/replace-servers-url.md).
+- Change your [OAuth2 token URL](../guides/change-token-url.md).
+
+## Preprocessors
+
+Preprocessors follow the same structure and operation as decorators, but they are run before the validation/linting step. This makes them brittle because the document may not be valid, and the extra processing step can cause performance impacts. We recommend looking for alternative approaches to preprocessing.
```

Comment:
```suggestion
Preprocessors follow the same structure and operation as decorators, but they are run before the validation/linting step. Running before the validation/linting step makes them brittle because the document may not be valid, and the extra processing step can cause performance impacts. We recommend looking for alternative approaches to preprocessing.
```